3. Import the required packages with their customary aliases.

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

4. Use the %matplotlib inline magic command so that your pltos show in the notebook without having to call plt.show() every time.

In [8]:
%matplotlib inline

5. Using the pandas read_csv() method, read the GDP dataset into your notebook as a DataFrame called gdp_df. Take a look at the first 6 rows.

In [9]:
gdp_df = pd.read_csv("../data/gdp_percapita.csv")

In [10]:
gdp_df.head(6)

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2019  2065.036398              NaN
1     Afghanistan  2018  2033.779002              NaN
2     Afghanistan  2017  2058.383832              NaN
3     Afghanistan  2016  2057.062164              NaN
4     Afghanistan  2015  2068.265904              NaN
5     Afghanistan  2014  2102.385234              NaN

Answer 5: The first six rows of data are presented above. The data includes 4 columns - Country or Area, Year, Value, and Value Footnotes. Upon further inspection in Visual Studio Code, the data set concludes on row 6869 before skipping a row and listing footnotes. As a revision, the dataframe will be amended to include a provision with nrows to ensure that the footnote details are not picked up as part of the dataset. 

In [11]:
gdp_df = pd.read_csv("../data/gdp_percapita.csv", nrows=6869)

6. Repeat for the internet use dataset. Call this DataFrame internet_df. Take a look at the first six rows.

In [12]:
internet_df = pd.read_csv("../data/internet_use.csv")

ParserError: Error tokenizing data. C error: Expected 4 fields in line 4675, saw 6


Given the ParserError located in line 4675 I checked the rows in Visual Studio Code and determined that the data ends in line 4496 before footnotes begin in 4498. I will rerun the code to create the dataframe using nrows in order to truncate the dataset to exclude the footnote details.

In [13]:
internet_df = pd.read_csv("../data/internet_use.csv", nrows=4496)

In [14]:
internet_df.head(6)

Country or Area  Year     Value  Value Footnotes
0     Afghanistan  2014  6.390000              NaN
1     Afghanistan  2013  5.900000              NaN
2     Afghanistan  2012  5.454545              NaN
3     Afghanistan  2011  5.000000              NaN
4     Afghanistan  2010  4.000000              NaN
5     Afghanistan  2009  3.550000              NaN

Answer 6. The information in the internet use DataFrame has the same basic format as the gdp DataFrame.

7. Look at the shape of each DataFrame - how many rows? How many columns?

In [15]:
gdp_df.shape

(6869, 4)

In [16]:
internet_df.shape

(4496, 4)

Answer 7. Both dataframes are set up with 4 columns. However, the gdp data frame is significantly larger with 6869 rows compared to the 4496 rows in the internet data frame.

8. Take a look at the datatypes for the columns in each DataFrame.

In [17]:
gdp_df.dtypes

Country or Area     object
Year                object
Value              float64
Value Footnotes    float64
dtype: object

In [18]:
internet_df.dtypes

Country or Area     object
Year                object
Value              float64
Value Footnotes    float64
dtype: object

Answer 8. In both dataframes, Country or Area and Year are object data types and Value and Value Footnotes are float64 data types.

9. Take a look at the last 10 rows of each DataFrame in turn.

In [19]:
gdp_df.tail(10)

Country or Area      Year        Value  Value Footnotes
6859        Zimbabwe      1998  3653.920016              NaN
6860        Zimbabwe      1997  3580.048793              NaN
6861        Zimbabwe      1996  3520.430146              NaN
6862        Zimbabwe      1995  3226.413930              NaN
6863        Zimbabwe      1994  3263.934978              NaN
6864        Zimbabwe      1993  3033.504852              NaN
6865        Zimbabwe      1992  3054.889178              NaN
6866        Zimbabwe      1991  3426.598094              NaN
6867        Zimbabwe      1990  3324.348171              NaN
6868   footnoteSeqID  Footnote          NaN              NaN

In [20]:
internet_df.tail(10)

Country or Area      Year     Value  Value Footnotes
4486        Zimbabwe      2001  0.799846              NaN
4487        Zimbabwe      2000  0.401434              NaN
4488        Zimbabwe      1999  0.161676              NaN
4489        Zimbabwe      1998  0.081648              NaN
4490        Zimbabwe      1997  0.033080              NaN
4491        Zimbabwe      1996  0.016790              NaN
4492        Zimbabwe      1995  0.007684              NaN
4493        Zimbabwe      1994  0.001739              NaN
4494        Zimbabwe      1990  0.000000              NaN
4495   footnoteSeqID  Footnote       NaN              NaN

Answer 9. The last 10 rows for each dataframe can be observed in Out[15] and Out[16]. Both final rows are related to footnotes rather than the data. This is due to a calculation error on my part as python starts counting at row 0 rather than row 1. This can be amended by recalculating the dataframe with a different value for nrows.

10. Drop the value footnotes column from both dataframes. Check that this worked as expected.

In [25]:
gdp_df.drop(gdp_df.columns[3], axis=1, inplace=True)
print(gdp_df)

     Country or Area      Year        Value
0        Afghanistan      2019  2065.036398
1        Afghanistan      2018  2033.779002
2        Afghanistan      2017  2058.383832
3        Afghanistan      2016  2057.062164
4        Afghanistan      2015  2068.265904
...              ...       ...          ...
6864        Zimbabwe      1993  3033.504852
6865        Zimbabwe      1992  3054.889178
6866        Zimbabwe      1991  3426.598094
6867        Zimbabwe      1990  3324.348171
6868   footnoteSeqID  Footnote          NaN

[6869 rows x 3 columns]


In [26]:
internet_df.drop(internet_df.columns[3], axis=1, inplace=True)
print(internet_df)

     Country or Area      Year     Value
0        Afghanistan      2014  6.390000
1        Afghanistan      2013  5.900000
2        Afghanistan      2012  5.454545
3        Afghanistan      2011  5.000000
4        Afghanistan      2010  4.000000
...              ...       ...       ...
4491        Zimbabwe      1996  0.016790
4492        Zimbabwe      1995  0.007684
4493        Zimbabwe      1994  0.001739
4494        Zimbabwe      1990  0.000000
4495   footnoteSeqID  Footnote       NaN

[4496 rows x 3 columns]


Answer 10. Both dataframes have been cleaned up to remove the Value Footnotes column through the use of the drop() command.

11. Change the columns for the GDP Per Capita DataFrame to 'Country', 'Year', and 'GDP_Per_Capita'

In [27]:
gdp_df = gdp_df.rename(columns = {'Country or Area' : 'Country', 'Value' : 'GDP_Per_Capita'})
print(gdp_df)

            Country      Year  GDP_Per_Capita
0       Afghanistan      2019     2065.036398
1       Afghanistan      2018     2033.779002
2       Afghanistan      2017     2058.383832
3       Afghanistan      2016     2057.062164
4       Afghanistan      2015     2068.265904
...             ...       ...             ...
6864       Zimbabwe      1993     3033.504852
6865       Zimbabwe      1992     3054.889178
6866       Zimbabwe      1991     3426.598094
6867       Zimbabwe      1990     3324.348171
6868  footnoteSeqID  Footnote             NaN

[6869 rows x 3 columns]


Answer 11. Country or Area and Value columns were renamed to Country and GDP_Per_Capita respectively by using the rename() function.

12. Change the columns for the Internet Users DataFrame to 'Country', 'Year', and 'Internet_Users_Pct'.

In [28]:
internet_df = internet_df.rename(columns = {'Country or Area' : 'Country', 'Value' : 'Internet_Users_Pct'})
print(internet_df)

            Country      Year  Internet_Users_Pct
0       Afghanistan      2014            6.390000
1       Afghanistan      2013            5.900000
2       Afghanistan      2012            5.454545
3       Afghanistan      2011            5.000000
4       Afghanistan      2010            4.000000
...             ...       ...                 ...
4491       Zimbabwe      1996            0.016790
4492       Zimbabwe      1995            0.007684
4493       Zimbabwe      1994            0.001739
4494       Zimbabwe      1990            0.000000
4495  footnoteSeqID  Footnote                 NaN

[4496 rows x 3 columns]


Answer 12. Country or Area and Value columns were renamed to Country and Internet_Users_Pct respectively by using the rename() function.

13. Merge the two DataFrames to one. Merge all rows from each of the two DataFrames. Call the new DataFrame gdp_and_internet_use.

In [30]:
gdp_and_internet_use = pd.merge(gdp_df,
                               internet_df[['Internet_Users_Pct']],
                               on='Country', 'Year')
print(gdp_and_internet_use)

SyntaxError: positional argument follows keyword argument (Temp/ipykernel_4132/506150456.py, line 3)

Will try to rerun using only one column for on.

In [31]:
gdp_and_internet_use = pd.merge(gdp_df,
                               internet_df[['Internet_Users_Pct']],
                               on='Country')
gdp_and_internet_use.head()

KeyError: 'Country'

Has resulted in duplicates.

In [34]:
gdp_and_internet_use = pd.merge(gdp_df, internet_df,
                               on = ['Country', 'Year'], 
                               how = 'outer')
gdp_and_internet_use.head()

Country  Year  GDP_Per_Capita  Internet_Users_Pct
0  Afghanistan  2019     2065.036398                 NaN
1  Afghanistan  2018     2033.779002                 NaN
2  Afghanistan  2017     2058.383832                 NaN
3  Afghanistan  2016     2057.062164                 NaN
4  Afghanistan  2015     2068.265904                 NaN

Looks to be in heading in the right direction, but need more of a preview to confirm that internet_users_pct column has merged properly.

In [35]:
gdp_and_internet_use = pd.merge(gdp_df, internet_df,
                               on = ['Country', 'Year'], 
                               how = 'outer')
gdp_and_internet_use.head(10)

Country  Year  GDP_Per_Capita  Internet_Users_Pct
0  Afghanistan  2019     2065.036398                 NaN
1  Afghanistan  2018     2033.779002                 NaN
2  Afghanistan  2017     2058.383832                 NaN
3  Afghanistan  2016     2057.062164                 NaN
4  Afghanistan  2015     2068.265904                 NaN
5  Afghanistan  2014     2102.385234            6.390000
6  Afghanistan  2013     2116.465454            5.900000
7  Afghanistan  2012     2075.491747            5.454545
8  Afghanistan  2011     1904.559799            5.000000
9  Afghanistan  2010     1957.029338            4.000000

Answer 13. DataFrames merged to create gdp_and_internet_use by merging on both Country and Year as part of an outer merge to ensure that the internet_users_pct column has transferred and is properly matched to the other columns.